components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Subscriber/src/Handlers/PubSubNetworkMessageBinaryHandler.cs

```diff
@@ -0,0 +1,91 @@
+// ------------------------------------------------------------
+//  Copyright (c) Microsoft Corporation.  All rights reserved.
+//  Licensed under the MIT License (MIT). See License.txt in the repo root for license information.
+// ------------------------------------------------------------
+
+namespace Microsoft.Azure.IIoT.OpcUa.Subscriber.Handlers {
+    using Microsoft.Azure.IIoT.Processor;
+    using Microsoft.Azure.IIoT.Processor.Models;
+    using Microsoft.Azure.IIoT.Hub;
+    using Newtonsoft.Json.Linq;
+    using Opc.Ua;
+    using Opc.Ua.PubSub;
+    using Serilog;
+    using System;
+    using System.IO;
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Text;
+    using System.Threading.Tasks;
+
+    /// <summary>
+    /// Publisher message handling
+    /// </summary>
+    public sealed class PubSubNetworkMessageBinaryHandler : IDeviceTelemetryHandler {
+
+        /// <inheritdoc/>
+        public string MessageSchema => Core.MessageSchemaTypes.NetworkMessageUadp;
+
+        /// <summary>
+        /// Create handler
+        /// </summary>
+        /// <param name="handlers"></param>
+        /// <param name="logger"></param>
+        public PubSubNetworkMessageBinaryHandler(IEnumerable<IMonitoredItemSampleProcessor> handlers, ILogger logger) {
+            _logger = logger ?? throw new ArgumentNullException(nameof(logger));
+            _handlers = handlers?.ToList() ?? throw new ArgumentNullException(nameof(handlers));
+        }
+
+        /// <inheritdoc/>
+        public async Task HandleAsync(string deviceId, string moduleId,
+            byte[] payload, IDictionary<string, string> properties, Func<Task> checkpoint) {
+            var json = Encoding.UTF8.GetString(payload);
+            var context = new ServiceMessageContext();
```

Comment:
>var context = new ServiceMessageContext(); [](start = 12, length = 42)

you could try also ServiceMessageContext.GlobalContext if you need a shared context some day with custom types

Reply:
thank's for the hint, I changed to ServiceMessageContext.GlobalContext

Reply:
welcome! Have you thought about supporting the complex types when the binary messages are decoded. Then it might be necessary to have a context provider which has the knowledge of the custom types. In theory this information could also be sent in MetaDataMessages, still then there needs to be a context provider which holds the type information specifically when converting to Json or cdm table data? Btw, nice job!

---
In reply to: [371513615](https://github.com/Azure/Industrial-IoT/pull/215#discussion_r371513615) [](ancestors = 371513615)

Reply:
I am not a fan of shared global context - it a) has a lock, and b) as the name says, completely jumble from all connected servers in it.   That said, if it is proven to be faster, maybe :-)

Reply:
Agreed, you save the constructor but you get a global lock in the max vars...better not use the global context here. For this application there should really be a context without locks..

Reply:
I think it is fast enough to create one for each encoder.

Reply:
agreed and reverted